src/confd/src/ietf-system.c

```diff
@@ -849,7 +853,110 @@ static char *change_get_user(struct sr_change *change)
 	return user;
 }
 
-static sr_error_t handle_sr_passwd_update(augeas *aug, sr_session_ctx_t *, struct sr_change *change)
+static int set_shell(const char *user, const char *shell)
+{
+	struct passwd *pw;
+	FILE *fp = NULL;
+	int fd = -1;
+
+	setpwent();
+
+	fp = fopen(_PATH_PASSWD "+", "w");
+	if (!fp) {
+		ERRNO("Failed opening %s+ for %s", _PATH_PASSWD, user);
+		goto fail;
+	}
+	fd = fileno(fp);
+	if (fd == -1 || fchown(fd, 0, 0) || fchmod(fd, 0644))
+		goto fail;
+
+	while ((pw = getpwent())) {
+		struct passwd upw;
+
+		if (!strcmp(pw->pw_name, user)) {
+			if (strcmp(pw->pw_shell, shell))
+				NOTE("Updating login shell for user %s to %s", user, shell);
+
+			upw = *pw;
+			upw.pw_shell = (char *)shell;
+			pw = &upw;
+		}
+
+		if (putpwent(pw, fp))
+			goto fail;
+	}
+
+	/* Ensure all calls are made, if either fails we bail */
+	if (fflush(fp) | fclose(fp) | rename(_PATH_PASSWD "+", _PATH_PASSWD)) {
```

Comment:
```suggestion
	if (fclose(fp) || rename(_PATH_PASSWD "+", _PATH_PASSWD)) {
```

I don't think the bitwise OR operator makes any promises about evaluation order. But in either case `fclose(3)` says...

> The `fclose()` function flushes the stream pointed to by `stream` (writing any buffered output data using `fflush(3)`) and closes the underlying file descriptor.

...so I think we could just elide the call and rely on logical OR instead.

Reply:
I added this was actually after studying the BusyBox sources.  With `||` the first failing call causes subsequent ones to be skipped.  With `|` we guarantee they are all called, hence the `ffush()` in case `fclose()` for some reason fails.

Reply:
Ok, but I don't think there's anything that says the compiler has to evaluate that from left to right, is there?

Also, if either the flush or close fails, don't we want to skip overwriting `/etc/passwd`? As I read it, we would be left without one now, no?

Reply:
Yeah good point, you're right!  I'll do the fixup as you suggested.